src/vmm/src/device_manager/persist.rs

```diff
@@ -119,6 +134,23 @@ impl<'a> Persist<'a> for MMIODeviceManager {
                 continue;
             }
 
+            #[cfg(target_arch = "aarch64")]
+            {
+                if *device_id == DeviceType::Serial.to_string()
+                    || *device_id == DeviceType::RTC.to_string()
+                {
+                    if states.legacy_devices.is_none() {
+                        states.legacy_devices = Some(Vec::new());
+                    }
+                    if let Some(legacy_devices) = states.legacy_devices.as_mut() {
```

Comment:
nit: all of this becomes much simpler if you drop `Option<Vec>` and use only `Vec`